zerver/actions/uploads.py

```diff
@@ -85,40 +88,56 @@ def do_delete_old_unclaimed_attachments(weeks_ago: int) -> None:
     # ArchiveAttachments in the same run; prevent warnings from the
     # backing store by only removing it from there once.
     already_removed = set()
     storage_paths = []
-    for attachment in old_unclaimed_attachments:
-        storage_paths.append(attachment.path_id)
-        image_row = ImageAttachment.objects.filter(path_id=attachment.path_id).first()
+    for attachment_path_id in attachments_path_ids:
+        storage_paths.append(attachment_path_id)
+        image_row = ImageAttachment.objects.filter(path_id=attachment_path_id).first()
         if image_row:
             for existing_thumbnail in image_row.thumbnail_metadata:
                 thumb = StoredThumbnailFormat(**existing_thumbnail)
                 storage_paths.append(get_image_thumbnail_path(image_row, thumb))
-            image_row.delete()
-        already_removed.add(attachment.path_id)
-        attachment.delete()
-        if len(storage_paths) >= DELETE_BATCH_SIZE:
-            delete_message_attachments(storage_paths[:DELETE_BATCH_SIZE])
-            storage_paths = storage_paths[DELETE_BATCH_SIZE:]
-    for archived_attachment in old_unclaimed_archived_attachments:
-        if archived_attachment.path_id not in already_removed:
-            storage_paths.append(archived_attachment.path_id)
-            image_row = ImageAttachment.objects.filter(path_id=archived_attachment.path_id).first()
+        already_removed.add(attachment_path_id)
+
+    for archived_attachment_path_id in archived_attachments_path_ids:
+        if archived_attachment_path_id not in already_removed:
+            storage_paths.append(archived_attachment_path_id)
+            image_row = ImageAttachment.objects.filter(path_id=archived_attachment_path_id).first()
             if image_row:  # nocoverage
                 for existing_thumbnail in image_row.thumbnail_metadata:
                     thumb = StoredThumbnailFormat(**existing_thumbnail)
                     storage_paths.append(get_image_thumbnail_path(image_row, thumb))
-                image_row.delete()
-        archived_attachment.delete()
-        if len(storage_paths) >= DELETE_BATCH_SIZE:
-            delete_message_attachments(storage_paths[:DELETE_BATCH_SIZE])
-            storage_paths = storage_paths[DELETE_BATCH_SIZE:]
-
-    if storage_paths:
-        delete_message_attachments(storage_paths)
+
+    # delete DELETE_BATCH_SIZE files one at a time, from storage
+    for batch_start_index in range(0, len(storage_paths), DELETE_BATCH_SIZE):
+        delete_message_attachments(
+            storage_paths[batch_start_index : batch_start_index + DELETE_BATCH_SIZE]
+        )
+
+    clear_old_unclaimed_attachments()
+
+
+def clear_old_unclaimed_attachments() -> None:
+    """delete old unclaimed attachments from database."""
+    Attachment.objects.filter(deleted=True).delete()
+    ArchivedAttachment.objects.filter(deleted=True).delete()
+    ImageAttachment.objects.filter(deleted=True).delete()
```

Comment:
See below, but these queries are inefficient without an index. You can generally use `.explain()` (printing it out) to check query plans in a management shell.

Reply:
**Testing after implementing partial index on `deleted=True`  and having `460` attachments** : 

We can see the index is now used when filtering by `deleted=True`, and the total estimated cost in the case of index `12` is less than that without index `24` 
 
```python
Attachment.objects.filter(deleted=False).explain() 
'Seq Scan on zerver_attachment  (cost=0.00..24.59 rows=457 width=122)\n  Filter: (NOT deleted)'

Attachment.objects.filter(deleted=True).explain()
'Index Scan using zerver_attachment_deleted_true_idx on zerver_attachment  (cost=0.13..12.32 rows=2 width=122)'

```